include/seqan3/io/alignment_file/format_sam.hpp

```diff
@@ -344,9 +619,406 @@ class alignment_file_format_sam
+            ranges::copy(stream_view | view::char_to<value_type_t<target_range_type>>,
+                         ranges::back_inserter(target));
+        else
+            std::ranges::next(std::ranges::begin(stream_view)); // skip '*'
+
+        assert(std::ranges::begin(stream_view) == std::ranges::end(stream_view)); // consumed everything
+    }
+
+    /*!\brief Reads arithmetic fields using std::from_chars.
+     * \tparam stream_view_type The type of the stream as a view.
+     * \tparam target_type      The type of value to parse from input; must model seqan3::Arithmetic.
+     *
+     * \param[in, out] stream_view  The stream view to iterate over.
+     * \param[in, out] target       The arithmetic value object to store the parsed value.
+     *
+     * \throws seqan3::parse_error if the character sequence in stream_view cannot be successfully converted to a value
+     *         of type target_type.
+     */
+    template <typename stream_view_type, Arithmetic target_type>
+    void read_field(stream_view_type && stream_view, target_type & target)
+    {
+        // unfortunately std::from_chars only accepts char const * so we need a buffer.
+        auto end = std::ranges::copy(stream_view, buffer.data()).out;
+        std::from_chars_result res = std::from_chars(buffer.begin(), end, target);
+
+        if (res.ec == std::errc::invalid_argument || res.ptr != end)
+            throw parse_error{std::string("[CORRUPTED SAM FILE] The string '") + std::string(buffer.begin(), end) +
+                                          "' could not be cast into type " +
+                                          detail::get_display_name_v<target_type>.string()};
+
+        if (res.ec == std::errc::result_out_of_range)
+            throw parse_error{std::string("[CORRUPTED SAM FILE] Casting '") + std::string(buffer.begin(), end) +
+                                          "' into type " + detail::get_display_name_v<target_type>.string() +
+                                          " would cause an overflow."};
+    }
+
+    /*!\brief Delegate parsing of std::optional types to parsing of the inner value type.
+     * \tparam stream_view_type     The type of the stream as a view.
+     * \tparam optional_value_type  The inner type of a the std::optional type of \p target.
+     *
+     * \param[in, out] stream_view  The stream view to iterate over.
+     * \param[in, out] target       The std::optional object to store the parsed value.
+     *
+     * \throws seqan3::parse_error if the character sequence in stream_view cannot be successfully converted to a value
+     *         of type target_type.
+     */
+    template <typename stream_view_type, typename optional_value_type>
+    void read_field(stream_view_type && stream_view, std::optional<optional_value_type> & target)
+    {
+        optional_value_type tmp;
+        read_field(std::forward<stream_view_type>(stream_view), tmp);
+        target = tmp;
```

Comment:
well, we so far only have an integer as an optional (ref_offset) which is read into a temporary value. For ref offset a 0 then indicates that the ref_offset needs to be unset which is handled on the higher level. This is only the case for ref_offset so it can't be pulled into this function. Additionally strings may never be empty in a SAM file because then they are set to `*` which I check on a higher level..

Reply:
Ok, so you read the value 0 for ref_offset, but then if it is zero, you unset the optional again at a higher level?

Reply:
almost. I read in zero into a temporary (I need the temporary because if I read in field::ALIGNMENT without field::REF_OFFSET I need to store the ref offset anyway for a correct alignment) and then test if the temporary is 0, if it is I do not set the actual ref_offset optional at all. See [here](https://github.com/seqan/seqan3/pull/632/files#diff-e1a35067660632d857b051141f685cdbR227)